core/src/services/gdrive/core.rs

```diff
@@ -190,29 +326,112 @@ impl GdriveCore {
             self.get_file_id_by_path(path).await?
         );
 
-        let req = self
-            .sign(Request::delete(&url))
+        let mut req = Request::delete(&url)
             .body(AsyncBody::Empty)
             .map_err(new_request_build_error)?;
 
+        req = self.sign(req);
+
+        self.client.send(req).await
+    }
+
+    pub async fn gdrive_upload_simple_request(
+        &self,
+        path: &str,
+        size: u64,
+        body: bytes::Bytes,
+    ) -> Result<Response<IncomingAsyncBody>> {
+        let parent = self.ensure_parent_path(path).await?;
+
+        let url = "https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart";
+
+        let file_name = path.split('/').filter(|&x| !x.is_empty()).last().unwrap();
+
+        let metadata = &json!({
+            "name": file_name,
+            "parents": [parent],
+        });
+
+        let req = Request::post(url).header("X-Upload-Content-Length", size);
+
+        let multipart = Multipart::new()
+            .part(
+                FormDataPart::new("metadata")
+                    .header(
+                        header::CONTENT_TYPE,
+                        "application/json; charset=UTF-8".parse().unwrap(),
+                    )
+                    .content(metadata.to_string()),
+            )
+            .part(
+                FormDataPart::new("file")
+                    .header(
+                        header::CONTENT_TYPE,
+                        "application/octet-stream".parse().unwrap(),
+                    )
+                    .content(body),
+            );
+
+        let mut req = multipart.apply(req)?;
+
+        req = self.sign(req);
+
         self.client.send(req).await
     }
 
-    fn sign(&self, mut req: Builder) -> Builder {
+    pub async fn gdrive_upload_overwrite_simple_request(
+        &self,
+        file_id: &str,
+        size: u64,
+        body: bytes::Bytes,
+    ) -> Result<Response<IncomingAsyncBody>> {
+        let url = format!(
+            "https://www.googleapis.com/upload/drive/v3/files/{}?uploadType=media",
+            file_id
+        );
+
+        let mut req = Request::patch(url)
+            .header(header::CONTENT_TYPE, "application/octet-stream")
+            .header(header::CONTENT_LENGTH, size)
+            .header("X-Upload-Content-Length", size)
+            .body(AsyncBody::Bytes(body))
+            .map_err(new_request_build_error)?;
+
+        req = self.sign(req);
+
+        self.client.send(req).await
+    }
+
+    fn sign<T>(&self, mut req: Request<T>) -> Request<T> {
         let auth_header_content = format!("Bearer {}", self.access_token);
-        req = req.header(header::AUTHORIZATION, auth_header_content);
+        req.headers_mut().insert(
+            header::AUTHORIZATION,
+            auth_header_content
+                .parse()
+                .expect("failed to parse auth header"),
```

Comment:
`access_token` is provided by users so it's possible to be invalid.

Thus, our sign API should be `sign<T>(&self, req: &mut Request<T>) -> Result<()>`.